### PR TITLE
Releasing v3.1.2

### PR DESCRIPTION
**Pull request by nkallima**
This release introduces the following changes/improvements:

- Expanding `validate.sh` to cover a few more concurrent objects.
- Increasing code coverage by `codecov`.
- Enhancing the API of `threadtools.h`.
- Code cleanup for the provided benchmark files.

### CODE REVIEW

File: libconcurrent/includes/threadtools.h [hunk 1]
```diff
@@ -25,17 +25,21 @@ int synchStartThreadsN(uint32_t nthreads, void *(*func)(void *), uint32_t uthrea
...
 void synchJoinThreadsN(uint32_t nthreads);
 
 /// @brief This function sets the CPU affinity of the running thread to cpu_id, where cpu_id
 /// should be a unique integer in {0, ..., N-1}, where N is the amount of available procissing cores.
 int synchThreadPin(int32_t cpu_id);
 
-/// @brief This function returns the CPU affinity of the running thread. More specifically, it returns
-/// a unique integer in {0, ..., N-1}, where N is the amount of available procissing cores.
+/// @brief This function returns the id of the running thread (posix or fiber). More specifically, it returns
+/// a unique integer in {0, ..., N-1}, where N is the amount of the running threads.
```
Comment (Spiros-N-Agathos) [inline libconcurrent/includes/threadtools.h:36]: So, if we have `3` Posix threads, and for each we have `4` fiber threads, this function will return `{0, ...., 11}`, right?
Maybe you can also add this simple example here.

File: libconcurrent/includes/uthreads.h [hunk 2]
```diff
@@ -27,4 +27,8 @@ int synchSpawnFiber(void *(*func)(void *), long arg);
 /// @brief Wait for all of the fibers that are running inside the current posix thread to finish their execution.
 void synchWaitForAllFibers(void);
 
+/// @brief This function returns an index of the running fiber in the current posix thread, i.e.
+/// an integer number between 0 and N-1, where N is the number of active fibers in the cirrent posix thread.
```
Comment (Spiros-N-Agathos) [inline libconcurrent/includes/uthreads.h:31]: `in the cirrent posix` --> `in the current posix`

File: libconcurrent/primitives/threadtools.c [hunk 3]
```diff
@@ -18,8 +18,8 @@ inline static void *uthreadWrapper(void *arg);
 inline static void *kthreadWrapper(void *arg);
 
 static __thread pthread_t *__threads;
-static __thread int32_t __thread_id = -1;
-static __thread int32_t __prefered_core = -1;
+static __thread int32_t __thread_id = 0;
```
Comment (Spiros-N-Agathos) [inline libconcurrent/primitives/threadtools.c:21]: why do we need these changes here?

**Review by Spiros-N-Agathos — CHANGES_REQUESTED** [review]

File: libconcurrent/concurrent/mcs.c [hunk 4]
```diff
@@ -16,8 +16,10 @@ void MCSLock(MCSLockStruct *l, MCSThreadState *thread_state, int pid) {
         prev->next = thread_state->MyNode;
         synchNonTSOFence();
     }
+
     while (thread_state->MyNode->locked == true) {
         synchResched();
+
```
[the remark this answers is not in the document]
Comment (nkallima) [reply libconcurrent/concurrent/mcs.c]: Fixed.

File: benchmarks/mcsbench.c [hunk 5]
```diff
@@ -55,12 +60,13 @@ int main(int argc, char *argv[]) {
...
     synchStartThreadsN(bench_args.nthreads, Execute, bench_args.fibers_per_thread);
-    synchJoinThreadsN(bench_args.nthreads - 1);
+    synchJoinThreadsN(bench_args.nthreads);
 
     printf("time: %d (ms)\tthroughput: %.2f (millions ops/sec)\t", (int)(d2 - d1), bench_args.runs * bench_args.nthreads / (1000.0 * (d2 - d1)));
     synchPrintStats(bench_args.nthreads, bench_args.total_runs);
 
 #ifdef DEBUG
+    fprintf(stderr, "DEBUG: Object state: %lu\n", debug_state);
     fprintf(stderr, "DEBUG: shared state: %f\n", object.state_f);
```
[the remark this answers is not in the document]
Comment (nkallima) [reply benchmarks/mcsbench.c]: Fixed.

File: libconcurrent/includes/threadtools.h (hunk 1 above)
Comment (nkallima) [reply libconcurrent/includes/threadtools.h]: Done.

File: libconcurrent/includes/uthreads.h (hunk 2 above)
Comment (nkallima) [reply libconcurrent/includes/uthreads.h]: Fixed

**Review by Spiros-N-Agathos — APPROVED** [review]

**Pull request merged**